staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -1251,6 +1252,7 @@ func (v *unstructuredSchemaCoercer) apply(u *unstructured.Unstructured) error {
 			// TODO: switch over pruning and coercing at the root to  schemaobjectmeta.Coerce too
 			structuralpruning.Prune(u.Object, v.structuralSchemas[gv.Version], false)
 		}
+		structuraldefaulting.PruneNulls(u.Object, v.structuralSchemas[gv.Version])
```

Comment:
why not inside the if statement?

Reply:
Now I am pretty sure we want this only for `!v.preserveUnknownFields`.

Reply:
> Now I am pretty sure we want this only for !v.preserveUnknownFields.

I'm trying to understand why. Clearly, unknownfields don't have a schema, so basically, they're always non-nullable non-defaulted, so all the nulls would be dropped. But I don't see why that's a problem, or consistent with what we're doing?

Reply:
> unknownfields don't have a schema, so basically, they're always non-nullable non-defaulted

I don't think that's true... with no schema, we allow any value, including null, so they are nullable non-defaulted

Reply:
Ah, that's a good point. My perspective has been that "nullable" is false by default, so a lack of schema would imply non-nullable.

Reply:
Does it mean that we wouldn't want to prune nulls for object where we conserve unknown fields? It sounds like we would want to preserve nulls if they don't have a schema.

Reply:
> It sounds like we would want to preserve nulls if they don't have a schema.

That's not what this code does today, but not because of that if/line.

Reply:
> Does it mean that we wouldn't want to prune nulls for object where we conserve unknown fields?

Only for unknown fields.

Reply:
Yes, good, that's what I expected.

Reply:
OK fixed!

Reply:
I consider the non-structural case as legacy we will never ever touch again. Hence, only change pruning of nulls in the if.

Reply:
As @apelisse found in our validation (he was 2s faster than me):
```
		if spec.PreserveUnknownFields == nil || *spec.PreserveUnknownFields == true {
			allErrs = append(allErrs, field.Invalid(fldPath.Child("preserveUnknownFields"), true, "must be false in order to use defaults in the schema"))
		}
```
We forbid defaulting for `preserveUnknownFields:true`. Hence, pruning some nulls without defaulting would be confusing.

Reply:
@sttts changed my mind by showing me this: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L252-L254